### ISSUE TITLE
User Defined Functions/ Table Valued Functions

### ISSUE THREAD

**Issue by sydsutton** [JordanMarr/SqlHydra#43]
Hello, 

I was wondering if you would be open to the idea of either adding and/or receiving a PR for User Defined Functions and Table Valued Functions? I use MS SQL Server, and would very much like to use SQL Hydra in conjunction with UDF's/TVF's. I might know someone who can help with the PR if that's the route you decide to go. 

Thanks so much.

**Comment by JordanMarr** [comment]
I assume this feature would exist within the `SqlHydra.Query` library. Would it involve type generation?
Can you give a code example of how you would imagine it would utilized within a query?

**Comment by sydsutton** [comment]
Yes, it would involve type generation. And I will get back to you as soon as possible with a code example.

**Comment by sydsutton** [comment]
This is my current code, just with the table that I'm referencing taking a parameter instead of taking no parameter. In MS SQL Server, the table being called would take the "calculation" that was passed in and do some math based on the calculation.
![image](https://user-images.githubusercontent.com/83406605/206583928-9939ce96-51c3-4b62-8fcc-b2fc5e843d52.png)

**Comment by JordanMarr** [comment]
What kind of support does `Microsoft.Data.SqlClient` provide for getting metadata on table valued functions?

**Comment by JordanMarr** [comment]
While `GetSchema` / `Procedures` does list UDFs, including a custom table function that I made, `GetSchema` / `Columns` doesn't list any of the returned columns. 

So there would need to be a metadata query that returns:
* The table valued function inputs (names, types)
* The table valued function output columns (names, types)

**Comment by sydsutton** [comment]
```sql 
-- To get Table Functions
SELECT * FROM INFORMATION_SCHEMA.ROUTINES WHERE ROUTINE_TYPE = 'FUNCTION' AND DATA_TYPE = 'TABLE'

-- To get Scalar functions, same as above except remove then DATA_TYPE filter
SELECT * FROM INFORMATION_SCHEMA.ROUTINES WHERE ROUTINE_TYPE = 'FUNCTION'

-- To get output columns
SELECT * FROM INFORMATION_SCHEMA.ROUTINE_COLUMNS -- Join on `TABLE_SCHEMA` and `TABLE_NAME`

-- To get input parameters
SELECT * FROM INFORMATION_SCHEMA.PARAMETERS -- Join on `SPECIFIC_SCHEMA` and `SPECIFIC_NAME`

**Comment by JordanMarr** [comment]
This seems doable. However, there is some refactoring in process that needs to be completed before this can be started.

**Comment by sydsutton** [comment]
Ok, that sounds good. Thank you for being open to the idea. Please let me know if there's anything else you need from me.

**Comment by RJSonnenberg** [comment]
@JordanMarr Is there any movement on this? I've run into an issue where I have to run both table and scalar functions and get the results. Would this solution be able to support scalar functions in addition to the table functions?

**Comment by JordanMarr** [comment]
It would be a nice feature to have for sure, but it’s not something I need and therefore i there is no movement. 
Currently, vanilla ado.net would probably be the easiest way to do it (unless you have a ton of functions, in which case I could understand why you would want this as a feature).

**Comment by sydsutton** [comment]
@RJSonnenberg I think I ended up using Dapper for this particular issue. I can send you more specifics in the morning if you need them.

**Comment by JordanMarr** [comment]
This feature looks like it could be challenging to implement because:
- different providers will likely have different queries to pull function names, inputs and outputs
- there are different kinds of functions, and some of them return custom types which is another feature that isn’t supported
- obviously the code generation piece would be difficult
- The Query library would need to somehow know that the generated functions were usable in the queries and that they should be spliced in (presumably by inspecting their path name)
